### PR TITLE
#158187265 Dockerize the application

### PR DESCRIPTION
**Pull request by makindemayowa**
#### What Does This PR Do?
- Setup docker for the application

#### Description Of Task To Be Completed
- Create `docker-compose.yml` for base application
- Create `Dockerfile` for the server, client and database
- Add setup scripts to the `Dockerfile`

#### Any Background Context You Want To Provide?
- To have a unified environment for the application

#### How can this be manually tested?
- Ensure docker is installed locally on machine
- Ensure these environment variables are added like so `DB_HOST=postgres, DB_NAME=a_socials, DB_USER=a_socials, DB_PASS=a_socials` `CLIENT_PORT=9000`
- To build and spin up the docker containers, run `docker-compose build`
- After building, run the containers with `make docker-build-client` or `make docker-build-server`

#### What are the relevant github issues?
N/A

#### Screenshot(s)

image

image

### CODE REVIEW

File: docker/server/entrypoint.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+set -e
+set -o pipefail # if any code doesn't return 0, exit the script
+
+
+echo "Waiting for postgres..."
```
Comment (andela-sjames) [inline docker/server/entrypoint.sh]: What check are we putting in place here?

File: docker/server/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.6.2
+
+ENV PYTHONUNBUFFERED 1
+
+# Setup Debian linux
+RUN export DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get -y install build-essential curl
+RUN curl -sL https://deb.nodesource.com/setup_8.x | bash -
+RUN apt-get install -y nodejs
+RUN nodejs -v && npm -v
```
Comment (andela-sjames) [inline docker/server/Dockerfile]: You might want to clean up after running apt-get `rm -rf /var/lib/apt/lists/*` 
reference: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      dockerfile: Dockerfile
+    ports:
+      - 5434:5432
+
+    volumes:
+      - andela_socials_db_data:/var/lib/postgresql/data
+      - andela_social_db_backup:/backups
+
+
+  andela_socials:
```
Comment (andela-sjames) [inline docker-compose.yml]: The naming is off, perhaps web service?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    container_name: andela_socials_server
+    build:
+      context: .
+      dockerfile: ./docker/server/Dockerfile
+    command:
+      - /start.sh
```
Comment (andela-sjames) [inline docker-compose.yml]: Better to split client and server for the sake of separation in the future, also for the sake of one container to one purpose ideology...

File: docker-compose.yml (hunk 3 above)
Comment (andela-sjames) [reply docker-compose.yml]: Is it just me, I can't seem to see the services shown on the pix in any of these files

File: docker-compose.yml (hunk 4 above)
Comment (tosmak16) [reply docker-compose.yml]: @andela-sjames Thanks. Initially, I suggested splitting client and server while discussing the acceptance criteria but an agreement was made as a team on using a single docker file for both server and client.

File: docker-compose.yml (hunk 3 above)
Comment (tosmak16) [reply docker-compose.yml]: @andela-sjames  Thanks I feel web service make sense we'll make necessary changes and we'll update the image to reflect the requested changes.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+docker-build:
+	docker-compose build
+docker-up:
+	docker-compose up
+
```
[the remark this answers is not in the document]
Comment (tosmak16) [reply Makefile]: @andela-sjames Thanks. we'll add them as well.

File: docker-compose-prod.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
+version: '3'
+
```
[the remark this answers is not in the document]
Comment (andela-sjames) [reply docker-compose-prod.yml]: While this holds in the land of DevOps within Andela Context I suppose the TTL needs this to replicate/know how certain changes will look like or behave in a near prod-like environment.

[the remark this answers is not in the document]
Comment (nyambati) [reply docker-compose-prod.yml]: Ok that makes sense,

File: docker-compose-dev.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      - andela_social_db_backup:/backups
+
+
+  web_server:
+    container_name: web_server
+    build:
+      context: .
+      dockerfile: ./docker/server/Dockerfile
+    environment:
+      - SLACK_BOT_TOKEN=${SLACK_BOT_TOKEN}
```
[the remark this answers is not in the document]
Comment (tosmak16) [reply docker-compose-dev.yml]: Okay...Thanks! We'll make necessary changes

File: .dockerignore [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# virtual environments
```
Comment (andela-sjames) [inline .dockerignore]: We may have to merge ideas from this PR into the other in general I say both PR's are good and merging the best of both worlds will be the best thing here.  Let's all have a sync tomorrow for the merger. This is the priority task IMHO!

Comment (nyambati) [inline .dockerignore]: I don't understand, why are we having two PRs on the same thing?

**Pull request closed without merging**